cmaes/_catcmawm.py

```diff
@@ -0,0 +1,1012 @@
+            / ((self._Nmi + 2) ** 2 + alpha_cov * self._mu_eff / 2),
+        )
+        assert (
+            self._c1 <= 1 - self._cmu
+        ), "Invalid learning rate for the rank-one update."
+        assert self._cmu <= 1 - self._c1, "Invalid learning rate for the rank-μ update."
+
+        min_alpha = (
+            0
+            if self._Nmi == 0
+            else min(
+                1 + self._c1 / self._cmu,
+                1 + (2 * mu_eff_minus) / (self._mu_eff + 2),
+                (1 - self._c1 - self._cmu) / (self._Nmi * self._cmu),
+            )
+        )
+
+        # TODO: Handle ranking ties when computing weights.
+        positive_sum = np.sum(weights_prime[weights_prime > 0])
+        negative_sum = np.sum(np.abs(weights_prime[weights_prime < 0]))
+        self._weights = np.where(
+            weights_prime >= 0,
+            1 / positive_sum * weights_prime,
+            min_alpha / negative_sum * weights_prime,
+        )
+
+        # generation number
+        self._g = 0
+
+        self._rng = np.random.RandomState(seed)
+
+        # --- initialization for each domain ---
+        if self._use_integer:
+            self._init_discretization(z_space)
+
+        if self._use_gaussian:
+            self._init_gaussian(x_space, mean, sigma, cov, population_size)
+
+        if self._use_categorical:
+            self._init_categorical(c_space, cat_param)
+
+    def _init_discretization(
+        self,
+        z_space: Optional[Sequence[Sequence[Union[int, float]]]],
+    ) -> None:
+        assert z_space is not None, "z_space must not be None for integer variables."
+
+        for i, row in enumerate(z_space):
+            if len(row) < 2:
+                raise ValueError(
+                    f"z_space must be a sequence of arrays with length >= 2. "
+                    f"Found length {len(row)} at index {i}: {row}"
+                )
+            if len(set(row)) < len(row):
+                raise ValueError(
+                    f"Elements in each array of z_space must be unique. "
+                    f"Found duplicate at index {i}: {row}"
+                )
+
+        # Pad the row with its maximum value to reach the maximum row length
+        max_length = max(len(row) for row in z_space)
+        self._z_space = np.array(
+            [
+                np.pad(
+                    np.array(sr),
+                    pad_width=(0, max_length - len(sr)),
+                    mode="constant",
+                    constant_values=(sr[-1]),
+                )
+                for row in z_space
+                for sr in [sorted(row)]
+            ]
+        )
+
+        # discretization thresholds
+        self._z_lim = (self._z_space[:, 1:] + self._z_space[:, :-1]) / 2
+
+        # margin value for integer variables
+        self._alpha = 1 - 0.73 ** (1 / (self._Nin + self._Nca))
+
+        # mutation rates for integer variables
+        self._pmut = (0.5 - 1e-10) * np.ones(self._Nin)
```

Comment:
To clarify that `1e-10` is used to avoid numerical errors, it may be better to use `_EPS` (e.g., https://github.com/CyberAgentAILab/cmaes/blob/main/cmaes/_cma.py#L11) if possible.
The value is currently 1e-8, but it is ok to decrease the value to 1e-10 (in my opinion).
This is also true in other places where ``1e-10`` is used (e.g., `_margin_correction` function).